core/src/main/java/org/bitcoinj/evolution/QuorumRotationState.java

```diff
@@ -864,4 +960,100 @@ public String toString() {
 
         return builder.toString();
     }
+
+    @Override
+    public void processDiff(@Nullable Peer peer, QuorumRotationInfo quorumRotationInfo, AbstractBlockChain headersChain, AbstractBlockChain blockChain, boolean isLoadingBootStrap) {
+        long newHeight = ((CoinbaseTx) quorumRotationInfo.getMnListDiffTip().coinBaseTx.getExtraPayloadObject()).getHeight();
+        if (peer != null)
+            peer.queueMasternodeListDownloadedListeners(MasternodeListDownloadedListener.Stage.Received, quorumRotationInfo.getMnListDiffTip());
+        Stopwatch watch = Stopwatch.createStarted();
+        Stopwatch watchMNList = Stopwatch.createUnstarted();
+        Stopwatch watchQuorums = Stopwatch.createUnstarted();
+        boolean isSyncingHeadersFirst = context.peerGroup != null && context.peerGroup.getSyncStage() == PeerGroup.SyncStage.MNLIST;
+        AbstractBlockChain chain = isSyncingHeadersFirst ? headersChain : blockChain;
+        log.info("processing quorumrotationinfo between : {} & {}; {}",
+                getMnListTip().getHeight(),
+                newHeight, quorumRotationInfo);
+
+        quorumRotationInfo.dump(getMnListTip().getHeight(), newHeight);
+
+        lock.lock();
+        try {
+            log.info(quorumRotationInfo.toString(chain));
+            setBlockChain(chain);
+            applyDiff(peer, headersChain, blockChain, quorumRotationInfo, isLoadingBootStrap);
+
+            log.info(this.toString());
+            unCache();
+            failedAttempts = 0;
+
+            if (!pendingBlocks.isEmpty()) {
+                StoredBlock thisBlock = pendingBlocks.get(0);
+                pendingBlocks.remove(0);
+                pendingBlocksMap.remove(thisBlock.getHeader().getHash());
+            } else log.warn("pendingBlocks is empty");
+
+            if (peer != null && isSyncingHeadersFirst)
+                peer.queueMasternodeListDownloadedListeners(MasternodeListDownloadedListener.Stage.Finished, quorumRotationInfo.getMnListDiffTip());
+
+        } catch (MasternodeListDiffException x) {
+            //we already have this mnlistdiff or doesn't match our current tipBlockHash
```

Comment:
handling of exceptions is very similar to `QuorumState.professDiff` which can allow for more refactoring